src/software/new_geom/linear_spline_test.cpp

```diff
@@ -0,0 +1,197 @@
+#include <gtest/gtest.h>
+
+#include "software/new_geom/linear_spline2d.h"
+
+TEST(LinearSplineTest, test_three_knot_value_at)
+{
+    std::vector<Point> points({Point(4, 18), Point(3, 0), Point(-7, -1)});
+    LinearSpline2d s(points);
+    EXPECT_EQ(s.getNumKnots(), points.size());
+    EXPECT_EQ(s.getValueAt(0.0), points[0]);
+    EXPECT_TRUE(s.getValueAt(0.1).isClose(Point(3.8, 14.4), 1e-9));
+    EXPECT_TRUE(s.getValueAt(0.5).isClose(points[1], 1e-9));
+    EXPECT_TRUE(s.getValueAt(0.6).isClose(Point(1, -0.2), 1e-8));
+    EXPECT_TRUE(s.getValueAt(0.9).isClose(Point(-5, -0.8), 1e-9));
+    EXPECT_EQ(s.getValueAt(1.0), points[2]);
+}
+
+TEST(LinearSplineTest, test_one_knot_value_at)
+{
+    std::vector<Point> points({Point(3, 7)});
+    LinearSpline2d s(points);
+    EXPECT_EQ(s.getNumKnots(), points.size());
+    EXPECT_EQ(s.getValueAt(0.0), points[0]);
+    EXPECT_EQ(s.getValueAt(0.1), points[0]);
+    EXPECT_EQ(s.getValueAt(0.5), points[0]);
+    EXPECT_EQ(s.getValueAt(0.6), points[0]);
+    EXPECT_EQ(s.getValueAt(0.9), points[0]);
+    EXPECT_EQ(s.getValueAt(1.0), points[0]);
+}
+
+TEST(LinearSplineTest, test_double_repeated_knot_value_at)
+{
+    std::vector<Point> points({Point(-3, 20), Point(-3, 20)});
+    LinearSpline2d s(points);
+    EXPECT_EQ(s.getNumKnots(), points.size());
+    EXPECT_EQ(s.getValueAt(0.0), points[0]);
+    EXPECT_EQ(s.getValueAt(0.1), points[0]);
+    EXPECT_EQ(s.getValueAt(0.5), points[0]);
+    EXPECT_EQ(s.getValueAt(0.6), points[0]);
+    EXPECT_EQ(s.getValueAt(0.9), points[0]);
+    EXPECT_EQ(s.getValueAt(1.0), points[0]);
+}
+
+TEST(LinearSplineTest, test_two_knot_value_at)
+{
+    std::vector<Point> points({Point(-3, -21), Point(4, 32)});
+    LinearSpline2d s(points);
+    EXPECT_EQ(s.getNumKnots(), points.size());
```

Comment:
nit: technically checking `NumKnots` is out of scope of this unit test (according to the name). Consider breaking it up